File: src/main/scala/com/github/bezsias/multimap/scala/MultiMap.scala
package com.github.bezsias.multimap.scala

trait MultiMap[K, V] {

  /** Returns the number of key-value pairs. */
  def size: Int

  def isEmpty: Boolean

  def contains(key: K): Boolean

  def contains(key: K, value: V): Boolean

  def keys: Set[K]

  def put(key: K, value: V)

  def putAll(key: K, values: List[V])

  def get(key: K): List[V]

  /** Removes all values associated with key. */
  def remove(key: K)

  /** Returns the count of removed elements */
  def remove(key: K, value: V)

  def clear()

}

File: src/main/java/com/github/bezsias/multimap/MultiMap.java
package com.github.bezsias.multimap;

import java.util.*;

/**
 * Map<K, List<V>>
 * Duplicate K-V pairs are allowed.
 * Values are ordered for a single key.
 */
public interface MultiMap<K, V> {

    /** Returns the number of key-value pairs. */
    int size();

    /** Lower limit on key size **/
    int minKeySize();

    /** Upper limit on key size **/
    int maxKeySize();

    boolean isEmpty();

    boolean contains(K key);

    boolean contains(K key, V value);

    Set<K> keys();

    void put(K key, V value);

    void putAll(K key, List<V> values);

    List<V> get(K key);

    /** Removes all values associated with a key. */
    void remove(K key);

    /** Returns the count of removed elements */
    void remove(K key, V value);

    void clear();
}
